### PR TITLE
Improving the Parser

### PR DESCRIPTION
**Pull request by schlunz**
Dear @otoolep,

In order to implement the `ecma404` support, I restructured the parser to make room for additional formats. 

Sincerely schlunz.

### CODE REVIEW

File: input/parser.go [hunk 1]
```diff
@@ -1,72 +1,66 @@
...
-	Priority  int    `json:"priority"`
-	Version   int    `json:"version"`
-	Timestamp string `json:"timestamp"`
-	Host      string `json:"host"`
-	App       string `json:"app"`
-	Pid       int    `json:"pid"`
-	MsgId     string `json:"msgid"`
-	Message   string `json:"message"`
+// HasFmt returns if the given format matches one of the possible formats.
+func HasFmt(format string) bool {
```
Comment (otoolep) [inline input/parser.go]: Can we rename this to `FormatSupported()`? Naming is hard. :-)

Comment (otoolep) [inline input/parser.go]: Or `ValidFormat`?

File: input/parser_test.go [hunk 2]
```diff
@@ -1,105 +1,91 @@
...
+	fail        bool
+	parserTests map[string]string
+}
 
-	tests := []struct {
-		message  string
-		expected RFC5424Message
-	}{
-		{
-			message:  "<134>1 2003-08-24T05:14:15.000003-07:00 ubuntu sshd 1999 - password accepted",
```
Comment (otoolep) [inline input/parser_test.go]: Where are all these tests now?

**Comment by otoolep** [comment]
OK, making sense to me now. It still seems a bunch of tests have been removed, and they need to go back back.

Can you also rebase with master? I merged your other PR.

**Comment by schlunz** [comment]
Thanks @otoolep for reviewing. I rebased and put the old test back into place.

Next, I will file a PR with the ecma404 support.

File: input/parser_test.go [hunk 3]
```diff
@@ -1,105 +1,209 @@
...
-			message:  "<1>0 2003-10-11T22:14:15.003Z test.com cron 65535 msgid1234 password accepted",
-			expected: RFC5424Message{Priority: 1, Version: 0, Timestamp: "2003-10-11T22:14:15.003Z", Host: "test.com", App: "cron", Pid: 65535, MsgId: "msgid1234", Message: "password accepted"},
-		},
-		{
-			message:  "<1>0 2003-08-24T05:14:15.000003-07:00 test.com cron 65535 - JVM NPE\nsome_file.java:48\n\tsome_other_file.java:902",
-			expected: RFC5424Message{Priority: 1, Version: 0, Timestamp: "2003-08-24T05:14:15.000003-07:00", Host: "test.com", App: "cron", Pid: 65535, MsgId: "-", Message: "JVM NPE\nsome_file.java:48\n\tsome_other_file.java:902"},
+var parserTests = ParserTestCases{
+	ParserTestCase{
+		fmt: "syslog",
+		tests: []string{
```
Comment (otoolep) [inline input/parser_test.go]: These tests are too difficult to maintain now, with the "input" and the "expected" disconnected like this. The code must use the older style.

File: input/parser_test.go [hunk 4]
```diff
@@ -1,105 +1,209 @@
...
+	},
+}
 
-	for i, tt := range tests {
-		m := p.Parse(tt.message)
-		if m == nil {
-			t.Fatalf("test %d: failed to parse: %s", i, tt.message)
+func Test_Formats(t *testing.T) {
+	var p *Parser
+	mismatched := func(rtrnd string, intnd string, intndA string) {
```
Comment (otoolep) [inline input/parser_test.go]: What does `intnd` mean? Intended?

Comment (schlunz) [inline input/parser_test.go]: It means `intended`.

**Comment by schlunz** [comment]
I will do the changes in the coming days.

**Comment by schlunz** [comment]
@otoolep Alright, I finally had time to fix the tests. :)

File: input/parser_test.go [hunk 5]
```diff
@@ -1,105 +1,214 @@
...
-			t.Logf("Message: %s (match: %v)", m.Message, m.Message == tt.expected.Message)
+		if !tt.fail && !reflect.DeepEqual(tt.expected, p.Result) {
+			t.Logf("%v", p.Result)
+			t.Logf("%v", tt.expected)
+			t.Error("\n\nParser result does not match expected result.\n")
 		}
 	}
 }
-
-func Test_FailedRFC5424Parsing(t *testing.T) {
```
Comment (otoolep) [inline input/parser_test.go]: Where is this test now?

File: input/parser_test.go [hunk 6]
```diff
@@ -1,105 +1,214 @@
...
- */
-
-func Test_SuccessfulRFC5424Parsing(t *testing.T) {
-	p := NewRFC5424Parser()
+func Test_Formats(t *testing.T) {
+	var p *Parser
+	mismatched := func(rtrnd string, intnd string, intndA string) {
+		if intndA != "" {
+			t.Fatalf("Parser format %v does not match the intended format %v.\n", rtrnd, intnd)
+		} else {
```
Comment (otoolep) [inline input/parser_test.go]: Go style dictates that there is no need for an `else` clause here. Just call the second `Fatalf` directly.

File: input/parser_test.go [hunk 7]
```diff
@@ -1,105 +1,214 @@
...
+			t.Fatalf("Parser format %v does not match the indended format %v (same as: %v).\n", rtrnd, intndA, intnd)
```
Comment (otoolep) [inline input/parser_test.go]: indended -> intended

**Comment by schlunz** [comment]
@otoolep I'm sorry. For some reason I forgot to include all tests, but now I fixed it.

**Comment by otoolep** [comment]
Thanks @schlunz

**Pull request merged**